datasync/chaindatafetcher/kafka/consumer.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2020 The klaytn Authors
+// This file is part of the klaytn library.
+//
+// The klaytn library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The klaytn library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the klaytn library. If not, see <http://www.gnu.org/licenses/>.
+
+package kafka
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/Shopify/sarama"
+)
+
+// TopicHandler is a handler function in order to consume published messages.
+type TopicHandler func(message *sarama.ConsumerMessage) error
+
+// Consumer is a reference structure to subscribe block or trace group produced by EN.
+type Consumer struct {
+	config   *KafkaConfig
+	group    sarama.ConsumerGroup
+	topics   []string
+	handlers map[string]TopicHandler
+}
+
+func NewConsumer(config *KafkaConfig, groupId string) (*Consumer, error) {
+	group, err := sarama.NewConsumerGroup(config.Brokers, groupId, config.SaramaConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &Consumer{
+		config:   config,
+		group:    group,
+		handlers: make(map[string]TopicHandler),
+	}, nil
+}
+
+// Close stops the ConsumerGroup and detaches any running sessions. It is required to call
+// this function before the object passes out of scope, as it will otherwise leak memory.
+func (c *Consumer) Close() error {
+	return c.group.Close()
+}
+
+// AddTopicAndHandler adds a topic associated the given event and its handler function to consume published messages of the topic.
+func (c *Consumer) AddTopicAndHandler(event string, handler TopicHandler) error {
+	if event != EventBlockGroup && event != EventTraceBroup {
+		return fmt.Errorf("the event name must be either 'blockgroup' or 'tracegroup'. given: %v", event)
+	}
+	topic := c.config.getTopicName(event)
+	c.topics = append(c.topics, topic)
+	c.handlers[topic] = handler
+	return nil
+}
+
+// Subscribe subscribes the registered topics with the handlers until the consumer is closed.
+func (c *Consumer) Subscribe(ctx context.Context) error {
```

Comment:
How about adding test that calls `Consumer.Subscribe` after `Consumer.Close` is called?

Reply:
I added more test cases. c53c490